cpp/src/spatial/point_in_polygon.cu

```diff
@@ -0,0 +1,233 @@
+        cudf::size_type poly_end = poly_offsets[poly_idx];
+        bool point_is_within = false;
+
+        // for each ring
+        for (cudf::size_type ring_idx = poly_begin; ring_idx < poly_end; ring_idx++)
+        {
+            cudf::size_type ring_begin = (ring_idx == 0) ? 0 : poly_ring_offsets[ring_idx - 1];
+            cudf::size_type ring_end = poly_ring_offsets[ring_idx];
+
+            // for each line segment
+            for (cudf::size_type point_idx = ring_begin; point_idx < ring_end - 1; point_idx++)
+            {
+                T ax = poly_points_x[point_idx];
+                T ay = poly_points_y[point_idx];
+                T bx = poly_points_x[point_idx + 1];
+                T by = poly_points_y[point_idx + 1];
+
+                auto y_between_ay_by = ay <= y && y < by; // is y in range [ay, by) when ay < by?
+                auto y_between_by_ay = by <= y && y < ay; // is y in range [by, ay) when by < ay?
+                auto y_in_bounds = y_between_ay_by || y_between_by_ay; // is y in range [by, ay]?
+                auto run  = bx - ax;
+                auto rise = by - ay;
+                auto rise_to_point =  y - ay;
+
+                if (y_in_bounds && x < (run / rise) * rise_to_point + ax) {
+                    point_is_within = not point_is_within;
+                }
+            }
+        }
+
+        hit_mask |= point_is_within << poly_idx;
+    }
+
+    result[idx] = hit_mask;
+}
+
+struct point_in_polygon_functor
+{
+    template <typename T>
+    static constexpr bool is_supported()
+    {
+        return std::is_floating_point<T>::value;
+    }
+
+    template <typename T, std::enable_if_t< !is_supported<T>() >* = nullptr, typename... Args>
+    std::unique_ptr<cudf::column> operator()(Args&& ...)
+    {
+        CUSPATIAL_FAIL("Non-floating point operation is not supported");
+    }
+
+    template <typename T, std::enable_if_t< is_supported<T>() >* = nullptr>
+    std::unique_ptr<cudf::column>
+    operator()(cudf::column_view const& test_points_x,
+               cudf::column_view const& test_points_y,
+               cudf::column_view const& poly_offsets,
+               cudf::column_view const& poly_ring_offsets,
+               cudf::column_view const& poly_points_x,
+               cudf::column_view const& poly_points_y,
+               rmm::mr::device_memory_resource* mr,
+               cudaStream_t stream)
+    {
+        auto size = test_points_y.size();
+        auto tid = cudf::experimental::type_to_id<cudf::size_type>();
+        auto type = cudf::data_type{ tid };
+        auto results = cudf::make_fixed_width_column(type,
+                                                     size,
+                                                     cudf::mask_state::UNALLOCATED,
+                                                     stream,
+                                                     mr);
+
+        if (results->size() == 0)
+        {
+            return results;
+        }
```

Comment:
cudf copying utils can simplify this a lot.
```suggestion
        if (test_points_y.is_empty()) return cudf::experimental::empty_like(test_points_y);
        auto results = cudf::expermental::detail::allocate_like(test_points_y, mask_allocation_policy::RETAIN, mr, stream);
```

Reply:
Unfortunately `empty_like` and `allocate_like` can't be used here because the result type differs from the input type.